### ISSUE TITLE
Completion show/commit behaviour is too intrusive for F#

### ISSUE THREAD

**Issue by dsyme** [mono/monodevelop#3798]
By default "Show completion list after character is typed" and "Complete with Space or Punctuation" are both on globally.

This is OK for C# but is resulting in quite intrusive intellisense completion for F#.  For example, C# doesn’t offer any completion at all in identifier positions.  The problem is that in F# it’s harder to determine identifier positions because…well… there are fewer identifiers.  

It looks like we suppress offering autocomplete after “let” and “type” but are showing it in other situations, e.g. when it commits to ‘System’ when you type ‘:’ in 

    type x = { Sys:

That case alone is noticeable, but we should enumerate further cases where intrusive completions have been offered  and see if there’s something we can do. We might need an adjustment in monodevelop to allow us to choose the characters that cause an eager commit.

> VS bug #563779

**Comment by nosami** [comment]
[From Visual F#](https://github.com/Microsoft/visualfsharp/blob/1e7037cba0119b0d53fc60c8cc0d7e3d7813ff17/vsintegration/src/FSharp.Editor/Completion/CompletionProvider.fs#L71)

```
let noCommitChars = [|' '; '='; ','; '.'; '<'; '>'; '('; ')'; '!'; ':'; '['; ']'; '|'|].ToImmutableArray()
```

**Comment by cartermp** [comment]
Criteria for when to trigger completion in the first place in VF#:

```fsharp
    static member ShouldTriggerCompletionAux(sourceText: SourceText, caretPosition: int, trigger: CompletionTriggerKind, getInfo: (unit -> DocumentId * string * string list)) =
        // Skip if we are at the start of a document
        if caretPosition = 0 then false
        // Skip if it was triggered by an operation other than insertion
        elif not (trigger = CompletionTriggerKind.Insertion) then false
        // Skip if we are not on a completion trigger
        else
            let triggerPosition = caretPosition - 1
            let triggerChar = sourceText.[triggerPosition]

            // do not trigger completion if it's not single dot, i.e. range expression
            if not Settings.IntelliSense.ShowAfterCharIsTyped && triggerPosition > 0 && sourceText.[triggerPosition - 1] = '.' then
                false
            else
                let documentId, filePath, defines = getInfo()
                CompletionUtils.shouldProvideCompletion(documentId, filePath, defines, sourceText, triggerPosition) &&
                (triggerChar = '.' || (Settings.IntelliSense.ShowAfterCharIsTyped && CompletionUtils.isStartingNewWord(sourceText, triggerPosition)))
```
and this whole file: https://github.com/Microsoft/visualfsharp/blob/master/vsintegration/src/FSharp.Editor/Completion/CompletionUtils.fs

**Comment by nosami** [comment]
Strange that the default set of characters that _should_ commit for C# is almost the same as the set of characters that shouldn't commit for F#.

```cs
const string commitChars = " <>()[]{}=+-*/%~&^|!.,;:?\"'";
```

https://github.com/mono/monodevelop/blob/4b8cf8c8e191ca65142d0b2afbc7131116818f65/main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.CodeCompletion/CompletionData.cs#L245

**Comment by nosami** [comment]
If I take the intersection of

```
const string commitChars = " <>()[]{}=+-*/%~&^|!.,;:?\"'"; 
```

and

```fs
let noCommitChars = [|' '; '='; ','; '.'; '<'; '>'; '('; ')'; '!'; ':'; '['; ']'; '|'|].ToImmutableArray()
```

I will have very few characters left that should auto-commit. Starting to think that we should never auto-commit.

**Comment by cartermp** [comment]
Part of why C# has a more robust "commit on this chars" set is because Roslyn has an incredible completion system via their provider architecture. At just about any point in any in-progress code, they have near full context on what is being typed and only show relevant completion there. We lack this for F# and thus are a bit more strict.

**Comment by xamarin-release-manager** [comment]
> [sync] [VS-22] Comment by **xamarinc**



Fixed in version 8.6.0.802 (master)



Pull Request #3887 merged by: Jason Imison  
 Author: mono  
 Commit: 3daa4434e6373e2a1ddf6df4fdcc0edcb3570dec (xamarin/monodevelop)
